### ISSUE TITLE
Hangs when writing credentials to standard output

### ISSUE THREAD

**Issue by iainelder** [linaro-its/aws2-wrap#51]
Thanks for this handy utility!

My desktop computer doesn't have enough memory to run some of my tests, so I use aws2-wrap to copy my SSO credentials to a bigger EC2 instance.

I do this by copying the generated credentials from my desktop to the EC2's default credentials file via SSH. I'm trying avoid polluting the my desktop's default credentials file as a by product.

The simplest way for me to do this is just to print the credentials to standard output and copy them to my clipboard to later paste into SSH.

But when I set /dev/stdout as the credentials file, aws2-wrap hangs until I cancel it with CTRL+C. After cancelling it I see this output:

```
$ aws2-wrap --generate --profile sandbox-mgmt --configfile /dev/null --credentialsfile /dev/stdout --outprofile default
^CTraceback (most recent call last):
  File "/home/isme/.local/bin/aws2-wrap", line 8, in <module>
    sys.exit(main())
  File "/home/isme/.local/pipx/venvs/aws2-wrap/lib/python3.8/site-packages/aws2wrap/__init__.py", line 475, in main
    process_cred_generation(
  File "/home/isme/.local/pipx/venvs/aws2-wrap/lib/python3.8/site-packages/aws2wrap/__init__.py", line 351, in process_cred_generation
    config.read(credentialsfile)
  File "/usr/lib/python3.8/configparser.py", line 697, in read
    self._read(fp, filename)
  File "/usr/lib/python3.8/configparser.py", line 1017, in _read
    for lineno, line in enumerate(fp, start=1):
KeyboardInterrupt
```

If I name a temporary file, then it works as expected. In this case I need to use cat to write the result to standard output.

```
$ aws2-wrap --generate --profile sandbox-mgmt --configfile /dev/null --credentialsfile creds --outprofile default
Credentials written to creds
Configuration written to /dev/null
The credentials will expire at 2022-01-16T07:33:27+00:00
$ cat creds
[default]
aws_access_key_id = ...
aws_secret_access_key = ...
aws_session_token = ...
```

How do I make aws2-wrap write to standard output directly? If that's not possible, could you change it to make it so?

It would just be a small quality-of-life improvement :-)

**Comment by pcolmer** [comment]
@iainelder The reason it doesn't work when you specify `/dev/stdout` as the value for `--credentialsfile` is because the script has to read the credentials file in first so that it can add the credentials to the file.

If all you want is the credentials to be displayed on your screen so that you can copy and paste them elsewhere, have you tried `aws2-wrap --profile sandbox-mgmt --export`?

**Comment by iainelder** [comment]
I had missed that option! I'll try it when I get back to my desk. Thanks for the quick feedback.

**Comment by iainelder** [comment]
The export option formats the credentials as bash environment variable export statements.

```
$ aws2-wrap --profile sandbox-mgmt --export
export AWS_ACCESS_KEY_ID=...
export AWS_SECRET_ACCESS_KEY=...
export AWS_SESSION_TOKEN=...
export AWS_DEFAULT_REGION=...
```

I want them formatted as a CLI config file section so that I can just copy and paste without any more manual reformatting.

That's why in my working example I write to a temporary file and print it. If aws2-wrap were about to write standard output, it would just be even more convenient.

> The reason it doesn't work when you specify /dev/stdout as the value for --credentialsfile is because the script has to read the credentials file in first so that it can add the credentials to the file.

That makes sense when the target is a normal file and already exists.

When the target is a new file, aws2-wrap must create it and write to it without reading anything, because there is no file to read from.

Standard out is a special case because the file already exists, but it can only be written and not read. Would you consider handling this case?

**Comment by pcolmer** [comment]
Hi @iainelder 

v1.2.4 adds `--generatestdout` which should do what you need it to do.

https://pypi.org/project/aws2-wrap/

**Comment by iainelder** [comment]
Hi, thanks for the quick response.

It's almost what I want, but not quite.

When I tried the generatestdout option, I couldn't get it to write the section header whose value is specified by outprofile.

```
$ aws2-wrap --generate --generatestdout --profile sandbox-mgmt --outprofile default
aws_access_key_id = ...
aws_secret_access_key = ...
aws_session_token = ...
```

I was expecting something like this:

```
$ aws2-wrap --generate --generatestdout --profile sandbox-mgmt --outprofile default
[default]
aws_access_key_id = ...
aws_secret_access_key = ...
aws_session_token = ...
```

That's what I would get if I wrote to a normal file and then catted the content.

The generatestdout option does nothing unless generate is also specified. That confused me until I read the source code to see how the options are parsed.

**Comment by pcolmer** [comment]
Hi @iainelder 

I've now released 1.2.6.

Instead of `--generatestdout` being a parameter to `--generate`, it is now a command option of its own so:

`aws2-wrap --generatestdout --profile sandbox-mgmt`

The profile header is now output and you don't need to specify it with `--outprofile` if you are wanting to use `default`.

**Comment by iainelder** [comment]
That's awesome! Thanks for the quick response again.

Now it behaves more or less like I would expect.

```
$ aws2-wrap --generatestdout --profile sandbox-mgmt
[profile default]
aws_access_key_id = ...
aws_secret_access_key = ...
aws_session_token = ...

$ aws2-wrap --generatestdout --profile sandbox-mgmt --outprofile sandbox-mgmt
[profile sandbox-mgmt]
aws_access_key_id = ...
aws_secret_access_key = ...
aws_session_token = ...
```

This almost seems nit-picky, but the behavior of the outprofile here is a bit surprising.

It seems like it would be less surprising to output the same profile name by default. When outprofile isn't specified in my example, I would expect to see "sandbox-mgmt" in the output.

I don't mind writing `--outprofile default` to change the name of the profile. It's just the name I was using in this example because I don't mind setting the default profile on an EC2 instance. I usually avoid setting the default profile anywhere else because of possible conflicts.

In any case, this is already a great little usability improvement for the tool. Thanks again.

**Comment by iainelder** [comment]
After copying and pasting the output to an EC2 instance I realised that the "profile" word in the section header should be removed.

For some reason the CLI credentials file has a different section header format from the CLI config file.

After I manually removed the "profile" word, the CLI was able to pick up the credentials pasted into ~/.aws/credentials.

So I think now I'm expecting output like this:

```
 $ aws2-wrap --generatestdout --profile sandbox-mgmt --outprofile default
[default]
aws_access_key_id = ...
aws_secret_access_key = ...
aws_session_token = ...
```
